File: src/main/kotlin/com/xs0/gqlktx/types/gql/GEnumType.java
package com.xs0.gqlktx.types.gql;

import com.xs0.gqlktx.QueryException;
import com.xs0.gqlktx.schema.builder.TypeKind;
import com.xs0.gqlktx.schema.intro.GqlIntroEnumValue;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public class GEnumType extends GValueType {
    private final Set<String> values;

    public GEnumType(String name, Set<String> values) {
        super(name);

        this.values = values;
    }

    @Override
    public TypeKind getKind() {
        return TypeKind.ENUM;
    }

    public Set<String> getValues() {
        return values;
    }

    @Override
    public void coerceValue(JsonObject raw, String key, JsonObject out) throws QueryException {
        try {
            out.put(key, check(raw.getString(key)));
        } catch (ClassCastException e) {
            throw new QueryException("Expected an enum value (String)");
        }
    }

    @Override
    public void coerceValue(JsonArray raw, int index, JsonArray out) throws QueryException {
        try {
            out.add(check(raw.getString(index)));
        } catch (ClassCastException e) {
            throw new QueryException("Expected an enum value (String)");
        }
    }

    private String check(String string) throws QueryException {
        if (string == null || values.contains(string))
            return string;

        throw new QueryException("Invalid enum value (" + string + "). The possibilities are: " + values);
    }

    @Override
    public void toString(StringBuilder sb) {
        sb.append("enum ").append(getName()).append(" {\n");
        for (String val : values)
            sb.append("  ").append(val).append("\n");
        sb.append("}\n");
    }

    public List<GqlIntroEnumValue> getValuesForIntrospection(boolean includeDeprecated) {
        ArrayList<GqlIntroEnumValue> res = new ArrayList<>();
        for (String value : values)
            res.add(new GqlIntroEnumValue(value));
        return res;
    }
}

File: src/main/kotlin/com/xs0/gqlktx/types/gql/GEnumType.kt
package com.xs0.gqlktx.types.gql

import com.xs0.gqlktx.QueryException
import com.xs0.gqlktx.dom.Value
import com.xs0.gqlktx.dom.ValueEnum
import com.xs0.gqlktx.dom.ValueString
import com.xs0.gqlktx.schema.builder.TypeKind
import com.xs0.gqlktx.schema.intro.GqlIntroEnumValue

import java.util.ArrayList

class GEnumType(name: String, val enumValues: List<GqlIntroEnumValue>, description: String?) : GValueType(name, description) {
    val stringValues = enumValues.map { it.name }

    override val kind: TypeKind
        get() = TypeKind.ENUM

    override fun coerceValue(raw: Value): ValueEnum {
        if (raw is ValueEnum) {
            check(raw.value)
            return raw
        }

        if (raw is ValueString) {
            check(raw.value)
            return ValueEnum(raw.value)
        }

        throw QueryException("Expected an enum value (String)")
    }

    private fun check(string: String): String {
        if (stringValues.contains(string))
            return string

        throw QueryException("Invalid enum value ($string). The possibilities are: $stringValues")
    }

    override fun toString(sb: StringBuilder) {
        sb.append("enum ").append(name).append(" {\n")
        for (`val` in stringValues)
            sb.append("  ").append(`val`).append("\n")
        sb.append("}\n")
    }

    fun getValuesForIntrospection(includeDeprecated: Boolean): List<GqlIntroEnumValue> {
        if (includeDeprecated) {
            return enumValues
        } else {
            return enumValues.filter { !it.isDeprecated }
        }
    }
}
